File: othello/play.py
"""Drive a self-play game with a chosen engine, printing each turn."""
from __future__ import annotations

from othello.core import Board, Move, winner
from othello.display import (
    format_board, format_score, player_name, move_name, BLACK_DOT_FG, RESET,
)
from othello.game import Engine, Depth, best_by_side
from othello.ai.alphabeta import AlphaBeta


def play_game(board: Board, engine: Engine | None = None,
              depth: Depth = None) -> Board:
    engine = AlphaBeta() if engine is None else engine
    last_move: Move | None = None
    turn = 1

    while not board.is_terminal():
        print()
        print(f"turn {turn}: {player_name(board.to_move)} to move")
        print(format_board(board, last_move=last_move))

        scored = engine.scores(board, depth)
        for move, score in scored:
            text = format_score(score)
            if score > 0:                       # black-favouring -> black-dot colour
                text = f"{BLACK_DOT_FG}{text}{RESET}"
            print(f"{move_name(move):>4} {text}")

        best_move = best_by_side(board, scored)
        print(f"best: {move_name(best_move)}")

        board = board.make_move(best_move)
        last_move = best_move
        turn += 1

    print()
    print("terminal")
    print(format_board(board, show_valid_moves=False, last_move=last_move))

    print(f"black: {board.black.bit_count()}")
    print(f"white: {board.white.bit_count()}")

    win = winner(board)
    print("winner: draw" if win is None else f"winner: {player_name(win)}")

    return board

File: rust/src/play.rs
//! Drive a self-play game with a chosen engine, printing each turn. Port of
//! `play.py`.

use crate::core::{winner, Board, Move};
use crate::display::{format_board, format_score, move_name, player_name, BLACK_DOT_FG, RESET};
use crate::game::{best_by_side, Depth, Engine};

pub fn play_game(mut board: Board, engine: &mut dyn Engine, depth: Depth) -> Board {
    let mut last_move: Option<Move> = None;
    let mut turn = 1;

    while !board.is_terminal() {
        println!();
        println!("turn {turn}: {} to move", player_name(board.to_move));
        println!("{}", format_board(&board, true, last_move));

        let scored = engine.scores(&board, depth).expect("non-terminal");
        for &(mv, score) in &scored {
            let text = format_score(score);
            let text = if score > 0 {
                format!("{BLACK_DOT_FG}{text}{RESET}") // black-favouring
            } else {
                text
            };
            println!("{:>4} {text}", move_name(mv));
        }

        let best = best_by_side(&board, &scored);
        println!("best: {}", move_name(best));

        board = board.make_move(best).expect("best move is legal");
        last_move = Some(best);
        turn += 1;
    }

    println!();
    println!("terminal");
    println!("{}", format_board(&board, false, last_move));
    println!("black: {}", board.black.count_ones());
    println!("white: {}", board.white.count_ones());
    match winner(&board) {
        None => println!("winner: draw"),
        Some(p) => println!("winner: {}", player_name(p)),
    }
    board
}
